crates/rpc/rpc-eth-types/src/revm_utils.rs

```diff
@@ -247,61 +275,67 @@ where
 fn apply_account_override<DB>(
     account: Address,
     account_override: AccountOverride,
-    db: &mut CacheDB<DB>,
+    db: &mut DB,
 ) -> EthResult<()>
 where
-    DB: DatabaseRef,
-    EthApiError: From<<DB as DatabaseRef>::Error>,
+    DB: Database + DatabaseCommit,
+    EthApiError: From<DB::Error>,
 {
-    // we need to fetch the account via the `DatabaseRef` to not update the state of the account,
-    // which is modified via `Database::basic_ref`
-    let mut account_info = db.basic_ref(account)?.unwrap_or_default();
+    let mut info = db.basic(account)?.unwrap_or_default();
 
     if let Some(nonce) = account_override.nonce {
-        account_info.nonce = nonce;
+        info.nonce = nonce;
     }
     if let Some(code) = account_override.code {
-        account_info.code = Some(
+        info.code = Some(
             Bytecode::new_raw_checked(code)
                 .map_err(|err| EthApiError::InvalidBytecode(err.to_string()))?,
         );
     }
     if let Some(balance) = account_override.balance {
-        account_info.balance = balance;
+        info.balance = balance;
     }
 
-    db.insert_account_info(account, account_info);
+    // Create a new account marked as touched
+    let mut acc =
+        revm::state::Account { info, status: AccountStatus::Touched, storage: HashMap::default() };
 
     // We ensure that not both state and state_diff are set.
     // If state is set, we must mark the account as "NewlyCreated", so that the old storage
     // isn't read from
-    match (account_override.state, account_override.state_diff) {
+    let storage_diff = match (account_override.state, account_override.state_diff) {
         (Some(_), Some(_)) => return Err(EthApiError::BothStateAndStateDiffInOverride(account)),
-        (None, None) => {
-            // nothing to do
-        }
-        (Some(new_account_state), None) => {
-            db.replace_account_storage(
+        (None, None) => None,
+        (Some(state), None) => {
+            // Destroy the account to ensure that its storage is cleared
+            db.commit(HashMap::from_iter([(
                 account,
-                new_account_state
-                    .into_iter()
-                    .map(|(slot, value)| {
-                        (U256::from_be_bytes(slot.0), U256::from_be_bytes(value.0))
-                    })
-                    .collect(),
-            )?;
-        }
-        (None, Some(account_state_diff)) => {
-            for (slot, value) in account_state_diff {
-                db.insert_account_storage(
-                    account,
-                    U256::from_be_bytes(slot.0),
-                    U256::from_be_bytes(value.0),
-                )?;
-            }
+                Account {
+                    status: AccountStatus::SelfDestructed | AccountStatus::Touched,
+                    ..Default::default()
+                },
+            )]));
+            Some(state)
```

Comment:
why does this work?

It is not obvious to me why we need to mark that account as destructed.